examples/chip-tool/commands/clusters/DataModelLogger.h

```diff
@@ -34,7 +52,122 @@ class DataModelLogger
+        value[kClusterIdKey]   = path.mClusterId;
+        value[kEndpointIdKey]  = path.mEndpointId;
+        value[kAttributeIdKey] = path.mAttributeId;
+
+        return LogError(value, status);
+    }
+
+    static CHIP_ERROR LogCommandAsJSON(const chip::app::ConcreteCommandPath & path, chip::TLV::TLVReader * data)
+    {
+        VerifyOrReturnError(mJSONDelegate != nullptr, CHIP_NO_ERROR);
+
+        Json::Value value;
+        value[kClusterIdKey]  = path.mClusterId;
+        value[kEndpointIdKey] = path.mEndpointId;
+        value[kCommandIdKey]  = path.mCommandId;
+
+        chip::TLV::TLVReader reader;
+        reader.Init(*data);
+        ReturnErrorOnFailure(chip::TlvToJson(reader, value));
+
+        auto valueStr = chip::JsonToString(value);
+        return mJSONDelegate->LogJSON(valueStr.c_str());
+    }
+
+    static CHIP_ERROR LogErrorAsJSON(const chip::app::ConcreteCommandPath & path, const chip::app::StatusIB & status)
+    {
+        VerifyOrReturnError(mJSONDelegate != nullptr, CHIP_NO_ERROR);
+
+        Json::Value value;
+        value[kClusterIdKey]  = path.mClusterId;
+        value[kEndpointIdKey] = path.mEndpointId;
+        value[kCommandIdKey]  = path.mCommandId;
+
+        return LogError(value, status);
+    }
+
+    static CHIP_ERROR LogEventAsJSON(const chip::app::EventHeader & header, chip::TLV::TLVReader * data)
+    {
+        VerifyOrReturnError(mJSONDelegate != nullptr, CHIP_NO_ERROR);
+
+        Json::Value value;
+        value[kClusterIdKey]  = header.mPath.mClusterId;
+        value[kEndpointIdKey] = header.mPath.mEndpointId;
+        value[kEventIdKey]    = header.mPath.mEventId;
+
+        chip::TLV::TLVReader reader;
+        reader.Init(*data);
+        ReturnErrorOnFailure(chip::TlvToJson(reader, value));
+
+        auto valueStr = chip::JsonToString(value);
+        return mJSONDelegate->LogJSON(valueStr.c_str());
+    }
+
+    static CHIP_ERROR LogErrorAsJSON(const chip::app::EventHeader & header, const chip::app::StatusIB & status)
+    {
+        VerifyOrReturnError(mJSONDelegate != nullptr, CHIP_NO_ERROR);
+
+        Json::Value value;
+        value[kClusterIdKey]  = header.mPath.mClusterId;
+        value[kEndpointIdKey] = header.mPath.mEndpointId;
+        value[kEventIdKey]    = header.mPath.mEventId;
+
+        return LogError(value, status);
+    }
+
+    static CHIP_ERROR LogErrorAsJSON(const CHIP_ERROR & error)
+    {
+        Json::Value value;
+        chip::app::StatusIB status;
+        status.InitFromChipError(error);
+        return LogError(value, status);
+    }
+
+    static void SetJSONDelegate(DataModelLoggerJSONDelegate * delegate) { mJSONDelegate = delegate; }
+
 private:
+    static CHIP_ERROR LogError(Json::Value & value, const chip::app::StatusIB & status)
+    {
+        if (status.mClusterStatus.HasValue())
+        {
+            auto statusValue          = status.mClusterStatus.Value();
+            value[kErrorIdKey]        = kGenericFailure;
```

Comment:
Can't we just get this for free via the "else" codepath, if we just always run that?